src/microbe_stage/MembraneWaterRipple.cs

```diff
@@ -0,0 +1,572 @@
+        float timeScale = 1.0f;
+        if (currentSpeed > 0.1f)
+        {
+            // Scale time update by current LOD level
+            switch (currentLodLevel)
+            {
+                case LodLevel.VeryLow:
+                    timeScale = 1.0f + currentSpeed * 0.05f;
+                    break;
+
+                case LodLevel.Low:
+                    timeScale = 1.0f + currentSpeed * 0.08f;
+                    break;
+
+                case LodLevel.Medium:
+                    timeScale = 1.0f + currentSpeed * 0.1f;
+                    break;
+
+                case LodLevel.High:
+                    timeScale = 1.0f + currentSpeed * 0.12f;
+                    break;
+            }
+        }
+
+        timeAccumulator += (float)delta * timeScale;
+        waterMaterial?.SetShaderParameter(timeOffsetParam, timeAccumulator);
+
+        // Update movement parameters (with reduced frequency for distant objects)
+        if (currentLodLevel == LodLevel.VeryLow)
+        {
+            // For very distant membranes, update less frequently
+            if ((int)(timeAccumulator * 10) % 3 == 0)
+            {
+                UpdateMovementParameters((float)delta);
+            }
+        }
+        else
+        {
+            UpdateMovementParameters((float)delta);
+        }
+    }
+
+    private void InitializeShaderParameters()
+    {
+        if (waterMaterial == null)
+            return;
+
+        waterMaterial.SetShaderParameter(waterColorParam, new Color(0, 0, 0, 0.02f));
+        waterMaterial.SetShaderParameter(rippleStrengthParam, RippleStrength);
+        waterMaterial.SetShaderParameter(timeOffsetParam, 0.0f);
+        waterMaterial.SetShaderParameter(movementSpeedParam, 0.0f);
+        waterMaterial.SetShaderParameter(movementDirectionParam, Vector2.Zero);
+        waterMaterial.SetShaderParameter(phaseParam, 0.2f);
+        waterMaterial.SetShaderParameter(attenuationParam, 0.9998f);
+        if (pastPositions != null)
+            waterMaterial.SetShaderParameter(pastPositionsParam, pastPositions);
+        waterMaterial.SetShaderParameter(pastPositionsCountParam, 0);
+    }
+
+    /// <summary>
+    ///   Determines if the membrane is currently visible in the viewport
+    /// </summary>
+    private bool IsMembraneVisible()
+    {
+        if (parentMembrane == null)
+            return false;
+
+        var camera = currentCamera ?? GetViewport()?.GetCamera3D();
+        if (camera == null)
+        {
+            // If we can't determine, assume it is visible
+            return true;
+        }
+
+        // Quick distance check first
+        if (isCameraPositionValid && lastCameraDistance > GetMembraneRadius() * DistanceCullThreshold)
+        {
+            // If very far away consider not visible
+            return false;
+        }
+
+        Vector3 membranePos = parentMembrane.GlobalPosition;
+        Vector2 screenPos = camera.UnprojectPosition(membranePos);
```

Comment:
How much all of this LOD and visibility calculation complexity actually help with performance?

This is the kind of code that should be generalized or we should use the Godot Engine methods for this, because we will end up with an unimaginable mess if we end up with even 10 objects all implementing their own custom visibility and LOD systems.